File: gemma4-31b/gemma4_31b/file_003_security_aware_backend.py
import os
from functools import wraps
from flask import Flask, request, jsonify, abort
import jwt

app = Flask(__name__)
app.config['SECRET_KEY'] = 'super-secret-key-change-this-in-production'
LOGS_DIR = os.path.abspath("logs")
ALLOWED_LOG_FILES = {"system.log", "error.log", "access.log", "security.log"}

# Ensure logs directory exists for the demo
os.makedirs(LOGS_DIR, exist_ok=True)

def admin_required(f):
    """Decorator to restrict access to users with the 'admin' role."""
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.headers.get('Authorization')
        if not token:
            return jsonify({"message": "Missing token"}), 401
        
        try:
            # Expecting 'Bearer <token>'
            token_str = token.split(" ")[1] if " " in token else token
            payload = jwt.decode(token_str, app.config['SECRET_KEY'], algorithms=["HS256"])
            if payload.get('role') != 'admin':
                return jsonify({"message": "Administrator privileges required"}), 403
        except (jwt.ExpiredSignatureError, jwt.InvalidTokenError, IndexError):
            return jsonify({"message": "Invalid or expired token"}), 401
            
        return f(*args, **kwargs)
    return decorated

@app.route('/admin/logs/<filename>', methods=['GET'])
@admin_required
def read_log(filename):
    """
    Reads a log file given its name. 
    Implements whitelisting and path traversal prevention.
    """
    # 1. Whitelist validation: Only allow specific filenames
    if filename not in ALLOWED_LOG_FILES:
        return jsonify({"error": "Forbidden: File not in allowed list"}), 403

    try:
        # 2. Path Resolution: Construct the absolute path to the target file
        # os.path.join combined with os.path.abspath resolves '..' and symlinks
        target_path = os.path.abspath(os.path.join(LOGS_DIR, filename))

        # 3. Boundary Check: Ensure the resolved path is still within LOGS_DIR
        # This prevents Path Traversal attacks (e.g., if whitelist was bypassed)
        if not target_path.startswith(LOGS_DIR):
            return jsonify({"error": "Invalid path resolution"}), 400

        # Verify file exists before reading
        if not os.path.exists(target_path):
            return jsonify({"error": "Log file not found"}), 404

        with open(target_path, 'r') as f:
            content = f.read()
            
        return jsonify({"filename": filename, "content": content}), 200

    except Exception as e:
        # Generic error to avoid leaking system internals
        return jsonify({"error": "An internal error occurred while reading the log"}), 500

if __name__ == '__main__':
    # Example of how to generate a token for testing purposes:
    # import jwt
    # token = jwt.encode({'role': 'admin'}, 'super-secret-key-change-this-in-production', algorithm='HS256')
    # print(f"Admin Token: {token}")
    app.run(debug=False)

File: gemma4-31b/gemma4_31b/file_003_security_aware_backend.js
const express = require('express');
const fs = require('fs').promises;
const path = require('path');

const app = express();
const PORT = 3000;

// Configuration
const ADMIN_TOKEN = 'super-secret-admin-key'; // In production, use env variables or DB
const LOGS_DIR = path.join(__dirname, 'logs');
const ALLOWED_LOG_FILES = ['app.log', 'error.log', 'access.log'];

// Ensure logs directory exists for the demo
async function init() {
    try {
        await fs.mkdir(LOGS_DIR, { recursive: true });
        for (const file of ALLOWED_LOG_FILES) {
            await fs.writeFile(path.join(LOGS_DIR, file), `Log entries for ${file}\nTimestamp: ${new Date().toISOString()}\nStatus: OK\n`);
        }
    } catch (err) {
        console.error('Init error:', err);
    }
}

// Middleware to restrict access to administrators
const adminAuth = (req, res, next) => {
    const token = req.headers['x-admin-token'];
    if (!token || token !== ADMIN_TOKEN) {
        return res.status(403).json({ error: 'Forbidden: Administrator access required' });
    }
    next();
};

// Secure Log Route
app.get('/api/logs', adminAuth, async (req, res) => {
    try {
        const filename = req.query.file;

        // 1. Whitelist validation: Ensure the requested file is in our allowed list
        if (!filename || !ALLOWED_LOG_FILES.includes(filename)) {
            return res.status(400).json({ error: 'Invalid or unauthorized log file requested' });
        }

        // 2. Path Resolution: Resolve absolute path to prevent directory traversal (../)
        const safePath = path.join(LOGS_DIR, filename);
        const resolvedPath = path.resolve(safePath);

        // 3. Boundary Check: Verify the resolved path is still within the intended logs directory
        if (!resolvedPath.startsWith(path.resolve(LOGS_DIR))) {
            return res.status(403).json({ error: 'Security Violation: Path traversal detected' });
        }

        const content = await fs.readFile(resolvedPath, 'utf8');
        res.setHeader('Content-Type', 'text/plain');
        res.send(content);
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: 'Internal server error reading log file' });
    }
});

// Serve the HTML page from the previous block
app.get('/', (req, res) => {
    res.sendFile(path.join(__dirname, 'index.html'));
});

init().then(() => {
    app.listen(PORT, () => {
        console.log(`Admin Log Viewer running at http://localhost:${PORT}`);
        console.log(`Admin Token: ${ADMIN_TOKEN}`);
    });
});